### PR TITLE
Introduce tests for MultiValueMap

### PR DESCRIPTION
**Pull request by midumitrescu**
imho improved tests for gh-25140, as solicited by @jhoeller 

As promised in gh-25159, I am submitting a second suggestion for Testing the exiting code of 

```
CollectionUtils.toMultiValueMap
```

The suggestion was to create for this MultiValueMap same tests as for 

```
LinkedMultiValueMap
```

present in
```
LinkedMultiValueMapTests
```

This, however, produced a very high code duplication, which I did not find very elegant. 

I think the better way is to use Jupiter properly, to instantiate various types of MultiValueMap and test the contract for each of them. 

Like this test code is not duplicated and the contracts are consistent with each other. 

This is just a suggestion. 

I am open for suggestions  in case this does not fit your expectations.

Kind regards

### CODE REVIEW

File: spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addIfAbsent("key", "value")),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addAll("key", exampleListOfValues())),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addAll(exampleMultiValueMap())),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.set("key", "value")),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.setAll(exampleHashMap())),
```
Comment (sbrannen) [inline spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java:172]: We tend to use `Collections.singletonMap` for such use cases.

```suggestion
						() -> asUnmodifiableMultiValueMap.setAll(Collections.singletonMap("key2", "key2.value1")),
```

File: spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		assertThat(objectUnderTest.getFirst("key")).isNull();
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void getFirst(MultiValueMap<String, String> objectUnderTest) {
+		List<String> values = new ArrayList<>(2);
+		values.add("value1");
+		values.add("value2");
+		objectUnderTest.put("key", values);
```
Comment (sbrannen) [inline spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java:103]: We tend to use `Arrays.asList` for such use cases.
```suggestion
		objectUnderTest.put("key", Arrays.asList("value1", "value2"));
```

File: spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+		o2.put("key1", Collections.singletonList("value1"));
+		assertThat(o2).isEqualTo(objectUnderTest);
+		assertThat(objectUnderTest).isEqualTo(o2);
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void canNotChangeAnUnmodifiableMultiValueMap(MultiValueMap<String, String> objectUnderTest) {
+		MultiValueMap<String, String> asUnmodifiableMultiValueMap = CollectionUtils.unmodifiableMultiValueMap(objectUnderTest);
+		Assertions.assertAll(
```
Comment (sbrannen) [inline spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java:160]: It is not permissible to use JUnit Jupiter's `Assertions`. Doing so will actually fail the build.

Thus, please switch to AssertJ assertions and be sure to execute `./gradlew check` to verify that you don't have any Checkstyle violations in this regard.

Note: if you want something similar to `assertAll`, you may wish to use `SoftAssertions` from AssertJ.

**Review by sbrannen — CHANGES_REQUESTED** [review]
I think the use of parameterized tests is a good idea.

I've requested a few minor changes and will review again once those have been made.

**Review by sbrannen — APPROVED** [review]

**Comment by midumitrescu** [comment]
@sbrannen i am on it :)

I need a couple of hours because of personal stuff and I will finish the PR afterwards.

**Comment by midumitrescu** [comment]
@sbrannen done

**Comment by snicoll** [comment]
Thanks @midumitrescu, I've enabled the tests that were disabled as the fix has been added in the meantime.

**Pull request closed without merging**